### ISSUE TITLE
Serde implementation should guard 128-bit integers

### ISSUE THREAD

**Issue by RReverser** [BurntSushi/rust-csv#158]
#### What version of the `csv` crate are you using?

`1.1.0`

#### Briefly describe the question, bug or feature request.

Code for 128-bit integers is included even on unsupported compilers / targets.

#### Include a complete program demonstrating a problem.

This leads not to an issue specific to code, but a compilation failure. Just include `csv` as a dependency and try to compile either for target without 128-bit integers support (like `wasm32-unknown-emscripten`) or using old Rust version (before 1.26).

#### What is the observed behavior of the code above?

```
   Compiling csv v1.1.0
error[E0407]: method `deserialize_u128` is not a member of trait `Deserializer`
   --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\deserializer.rs:324:9
    |
324 | /         fn $method<V: Visitor<'de>>(
325 | |             self,
326 | |             visitor: V,
327 | |         ) -> Result<V::Value, Self::Error> {
...   |
335 | |             visitor.$visit(num.map_err(|err| self.error(DEK::ParseInt(err)))?)
336 | |         }
    | |_________^ not a member of trait `Deserializer`
...
367 |       deserialize_int!(deserialize_u128, visit_u128, u128);
    |       ----------------------------------------------------- in this macro invocation

error[E0407]: method `deserialize_i128` is not a member of trait `Deserializer`
   --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\deserializer.rs:324:9
    |
324 | /         fn $method<V: Visitor<'de>>(
325 | |             self,
326 | |             visitor: V,
327 | |         ) -> Result<V::Value, Self::Error> {
...   |
335 | |             visitor.$visit(num.map_err(|err| self.error(DEK::ParseInt(err)))?)
336 | |         }
    | |_________^ not a member of trait `Deserializer`
...
372 |       deserialize_int!(deserialize_i128, visit_i128, i128);
    |       ----------------------------------------------------- in this macro invocation

error[E0407]: method `serialize_i128` is not a member of trait `Serializer`
  --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\serializer.rs:68:5
   |
68 | /     fn serialize_i128(self, v: i128) -> Result<Self::Ok, Self::Error> {
69 | |         self.collect_str(&v)
70 | |     }
   | |_____^ not a member of trait `Serializer`

error[E0407]: method `serialize_u128` is not a member of trait `Serializer`
  --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\serializer.rs:92:5
   |
92 | /     fn serialize_u128(self, v: u128) -> Result<Self::Ok, Self::Error> {
93 | |         self.collect_str(&v)
94 | |     }
   | |_____^ not a member of trait `Serializer`

error[E0407]: method `serialize_i128` is not a member of trait `Serializer`
   --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\serializer.rs:520:5
    |
520 | /     fn serialize_i128(self, v: i128) -> Result<Self::Ok, Self::Error> {
521 | |         self.handle_scalar(v)
522 | |     }
    | |_____^ not a member of trait `Serializer`

error[E0407]: method `serialize_u128` is not a member of trait `Serializer`
   --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\serializer.rs:540:5
    |
540 | /     fn serialize_u128(self, v: u128) -> Result<Self::Ok, Self::Error> {
541 | |         self.handle_scalar(v)
542 | |     }
    | |_____^ not a member of trait `Serializer`

error[E0599]: no method named `visit_u128` found for type `V` in the current scope
   --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\deserializer.rs:219:21
    |
219 |             visitor.visit_u128(n)
    |                     ^^^^^^^^^^ help: did you mean: `visit_u16`

error[E0599]: no method named `visit_i128` found for type `V` in the current scope
   --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\deserializer.rs:221:21
    |
221 |             visitor.visit_i128(n)
    |                     ^^^^^^^^^^ help: did you mean: `visit_i16`

error[E0599]: no method named `visit_u128` found for type `V` in the current scope
   --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\deserializer.rs:309:21
    |
309 |             visitor.visit_u128(n)
    |                     ^^^^^^^^^^ help: did you mean: `visit_u16`

error[E0599]: no method named `visit_i128` found for type `V` in the current scope
   --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\deserializer.rs:311:21
    |
311 |             visitor.visit_i128(n)
    |                     ^^^^^^^^^^ help: did you mean: `visit_i16`

error[E0599]: no method named `visit_u128` found for type `V` in the current scope
   --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\deserializer.rs:367:40
    |
367 |     deserialize_int!(deserialize_u128, visit_u128, u128);
    |                                        ^^^^^^^^^^ help: did you mean: `visit_u16`

error[E0599]: no method named `visit_i128` found for type `V` in the current scope
   --> C:\Users\Ingvar\.cargo\registry\src\github.com-1ecc6299db9ec823\csv-1.1.0\src\deserializer.rs:372:40
    |
372 |     deserialize_int!(deserialize_i128, visit_i128, i128);
    |                                        ^^^^^^^^^^ help: did you mean: `visit_i16`

error: aborting due to 12 previous errors
```

#### What is the expected or desired behavior of the code above?

It compiles with no errors.

---

To fix this, all usages of 128-bit methods in Serde code should be wrapped into [`serde_if_integer128!`](https://docs.serde.rs/serde/macro.serde_if_integer128.html) macro provided by Serde itself specifically for these compatibility purposes.

**Comment by RReverser** [comment]
cc @ian-hamlin

**Comment by BurntSushi** [comment]
> or using old Rust version (before 1.26).

csv 1.1 [bumped its MSRV](https://github.com/BurntSushi/rust-csv/blob/eafc208ccef047d9fe81452f16588c3d48fd6a5d/.travis.yml#L4), so this particular aspect is a non-issue.

> try to compile either for target without 128-bit integers support (like `wasm32-unknown-emscripten`)

The serde docs for `serde_if_integer128!` say:

> Data formats that require a minimum Rust compiler version of at least 1.26 do not need to bother with this macro and may assume support for 128-bit integers.

However, looking at serde's build.rs, it looks the docs are out of date, and 128-bit integer support is gated on both the Rust version and the target: https://github.com/serde-rs/serde/blob/1c1eecabc060738b3551dd9133389048fe34016c/serde/build.rs#L51-L58

I've filed a PR upstream to see if my understanding is right, and if so, update the docs: https://github.com/serde-rs/serde/pull/1561

---

Out of curiosity, in what context are you using csv on the emscripten target?

**Comment by BurntSushi** [comment]
Also, I have to say that I am a bit baffled that 128-bit support isn't provided for every target. This seems like a pretty inconvenient footgun. This effectively means every use of 128-bit integers needs to be gated behind something. Is this actually true, or is Serde being conservative here? In particular, byteorder has automatically enabled 128-bit support for integers for quite a while now, and I haven't heard of any problems there. If csv is being used on emscripten, then byteorder almost certainly is too.

**Comment by BurntSushi** [comment]
It looks like LLVM complains if one tries to use 128-bit integers on emscripten targets: https://github.com/serde-rs/serde/pull/1365 --- Is that still the case today?

**Comment by RReverser** [comment]
> It looks like LLVM complains if one tries to use 128-bit integers on emscripten targets

Unfortunately, yes, and this is what macro would help with (at least when using the current crate in dependencies).

**Comment by RReverser** [comment]
> Out of curiosity, in what context are you using csv on the emscripten target?

I'm trying to see if I can get `criterion` to run directly for WASM (via Emscripten, because that's the easiest way to get all the necessary glue), and for now looking into compilation failures that prevent it.

**Comment by BurntSushi** [comment]
Cool. All righty, this should be fixed in `csv 1.1.1` on crates.io. I tested this by running `cross build --target wasm32-unknown-emscripten`, which reproduced your errors. I however wasn't able to run `cross test --target wasm32-unknown-emscripten --lib` successfully due to what appear to be unrelated errors in cross's infrastructure. But hopefully this is good enough!

**Comment by RReverser** [comment]
@BurntSushi Thank you!
